### PR TITLE
[terraformer-v2] Add Pod e2e test

### LINKED ISSUE #10: Stand-Alone Integration Tests
This component lacks integration tests. Should we add them, so that integration in the Gardener becomes safer (in the context of the planned full output qualification and continuous deployment).

However, we actually plan to replace Terraform eventually with a native implementation (putting to icebox until clarification).

### PR DESCRIPTION
**Pull request by timebertt**
/area quality
/kind test

**What this PR does / why we need it**:
This PR adds an e2e test, that is executed by the test-machinery. It functions similarly to some infrastructure integration tests in our provider extensions.

It uses an existing cluster (pointed to by the `KUBECONFIG` env var) and deploys a terraformer `apply` Pod, that will create some lightweight resource (ec2 keypair) on AWS.
The test validates, that the resource was created on AWS using the AWS go-sdk and that the state ConfigMap has been updated accordingly. 
After that, the test deploys a terraformer `destroy` Pod and validates again the changes on AWS and the state ConfigMap.

**Which issue(s) this PR fixes**:
Fixes #10 

**Special notes for your reviewer**:
/invite @gardener/test-infra-maintainers 

**Release note**:

```noteworthy developer
You can now run a small e2e test, that creates some lightweight resource on AWS via terraform, by executing `make test-e2e`.
```

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -143,6 +147,15 @@ format:
 test:
 	@$(REPO_ROOT)/vendor/github.com/gardener/gardener/hack/test.sh ./cmd/... ./pkg/... ./test/e2e/binary/...
 
+.PHONY: test-e2e
+test-e2e:
+	@go test -timeout=0 -mod=vendor -ldflags $(LD_FLAGS) ./test/e2e/pod \
+		--v -ginkgo.v -ginkgo.progress \
+		--kubeconfig=${KUBECONFIG} \
+		--access-key-id='$(shell cat $(ACCESS_KEY_ID_FILE))' \
```
Comment (schrodit) [inline Makefile:155]: the testmachinery serves the access-key-id as env var with the name `ACCESS_KEY_ID`

File: Makefile [hunk 2]
```diff
@@ -143,6 +147,15 @@ format:
...
+		--secret-access-key='$(shell cat $(SECRET_ACCESS_KEY_FILE))' \
```
Comment (schrodit) [inline Makefile:156]: and the access key as env var with the name `SECRET_ACCESS_KEY`

**Review by schrodit — CHANGES_REQUESTED** [review]

**Comment by schrodit** [comment]
> Testrun: [e2e-w6m98](https://tm.gardener.cloud/testrun/default/e2e-w6m98)
> Workflow: e2e-w6m98-wf
> Phase: Failed
> 
> +--------------+--------------+--------+----------+
> |     NAME     |     STEP     | PHASE  | DURATION |
> +--------------+--------------+--------+----------+
> | pod-e2e-test | pod-e2e-test | Failed | 21m48s   |
> +--------------+--------------+--------+----------+

The pods is pending so I think there could be an issue with the version.

**Comment by timebertt** [comment]
Yes, already on it

**Comment by timebertt** [comment]
Update:
The test on its own is working fine (locally). The problem here is, that the `make` process in the test run needs to know, which commit / image tag it should test. It seems like the test-machinery is currently not mounting the `.git` directory, so we can't detect this automatically via `git rev-parse`.
@dguendisch and @schrodit are checking how this can be solved. Thanks!

**Comment by dguendisch** [comment]
I had a quick look in the workflow, @schrodit please correct me if I'm wrong:
the repo is currently synced to a file `$TM_REPO_PATH/<org>-<repo>-<commit-sha>` (where `TM_REPO_PATH` is an env variable). So before we add some more "stable" API way (i.e. some env `GIT_COMMIT`) you could parse this filename for the time being as a work-around.

**Comment by timebertt** [comment]
Had another chat with @schrodit and @dguendisch.
We agreed on injecting the commit sha and the tag (in case it's a release), which shall be tested, into the test runs via env vars.
Once this is done, I will adapt the PR to make use of that.

**Comment by timebertt** [comment]
/draft

Should be based on https://github.com/gardener/terraformer/pull/53 and waiting for test-machinery dependencies.

**Comment by timebertt** [comment]
Update:
- added the e2e test to release job
- switched to a chart for the test run to reuse it for bot triggered tests and release job

**Review by rfranzke — APPROVED** [review]
/lgtm

**Comment by timebertt** [comment]
Thank you very much, @schrodit! 🎉

File: .ci/test-e2e.sh [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+set -o nounset
+set -o pipefail
+
+cd "$(dirname $0)/.."
+
+version="$(cat ./VERSION)"
+
+mkdir -p /tm
+/cc/utils/cli.py config attribute --cfg-type kubernetes --cfg-name testmachinery --key kubeconfig > /tm/kubeconfig
+/testrunner run \
```
Comment (schrodit) [inline .ci/test-e2e.sh:13]: @dguendisch can you also have a look and double check if I missed soemthing

File: .test-defs/pod-e2e.yaml [hunk 4]
```diff
@@ -0,0 +1,17 @@
+kind: TestDefinition
+metadata:
+  name: pod-e2e-test
+spec:
+  owner: gardener-oq@listserv.sap.com
+  description: Terraformer Pod E2E test
+  activeDeadlineSeconds: 1800
```
Comment (schrodit) [inline .test-defs/pod-e2e.yaml:7]: The teststep timeout should be little bit higher that the timeout that is defined in the go test.

Comment (timebertt) [reply .test-defs/pod-e2e.yaml]: Yep, done. Also, decreased the overall timeouts.

**Review by rfranzke — APPROVED** [review]

File: .ci/test-e2e.sh [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+    --tm-kubeconfig-path=/tm/kubeconfig \
+    --timeout=1800 \
+    --testruns-chart-path=.ci/testruns/default-testrun \
```
Comment (dguendisch) [inline .ci/test-e2e.sh:16]: I think the path needs to be a fs path so it would be probably: `.ci/testruns/default` ?

**Review by dguendisch — CHANGES_REQUESTED** [review]

File: .ci/test-e2e.sh (hunk 5 above)
Comment (timebertt) [reply .ci/test-e2e.sh]: Good catch!

**Review by dguendisch — APPROVED** [review]

**Review by schrodit — APPROVED** [review]

**Pull request merged**